rust/src/origin.rs

```diff
@@ -6,56 +6,236 @@
+    }
+    kf.set_value(group.as_ref(), k.as_ref(), v.as_str())
+}
+
+fn set_sha256_nevra_pkgs(
+    kf: &glib::KeyFile,
+    group: &str,
+    k: &str,
+    pkgs: &BTreeMap<String, String>,
+) {
+    let pkgs: Vec<_> = pkgs
+        .iter()
+        .map(|(nevra, sha256)| format!("{}:{}", sha256, nevra))
+        .collect();
+    let pkgs = pkgs.iter().map(|s| s.as_str());
+    kf_set_string_list(&kf, group, k, pkgs)
+}
+
+/// Convert a treefile to an origin file.
+#[context("Parsing treefile origin")]
+fn treefile_to_origin_inner(tf: &Treefile) -> Result<glib::KeyFile> {
+    let tf = &tf.parsed;
+    let kf = glib::KeyFile::new();
+
+    // refspec (note special handling right now for layering)
+    let deriving = tf.packages.is_some() || tf.derive.packages_local.is_some();
+    if let Some(r) = tf.derive.from.as_deref() {
+        let k = if deriving { "baserefspec" } else { "refspec" };
+        kf.set_string(ORIGIN, k, r)
+    };
+
+    // Packages
+    if let Some(pkgs) = tf.packages.as_deref() {
+        let pkgs = pkgs.iter().map(|s| s.as_str());
+        kf_set_string_list(&kf, PACKAGES, "requested", pkgs)
+    }
+    if let Some(pkgs) = tf.derive.packages_local.as_ref() {
+        set_sha256_nevra_pkgs(&kf, PACKAGES, "requested-local", pkgs)
+    }
+    if let Some(pkgs) = tf.derive.override_remove.as_deref() {
+        let pkgs = pkgs.iter().map(|s| s.as_str());
+        kf_set_string_list(&kf, OVERRIDES, "remove", pkgs)
+    }
+    if let Some(pkgs) = tf.derive.override_replace_local.as_ref() {
+        set_sha256_nevra_pkgs(&kf, OVERRIDES, "replace-local", pkgs)
+    }
+
+    // Initramfs bits
+    if let Some(initramfs) = tf.derive.initramfs.as_ref() {
+        if initramfs.regenerate {
+            kf.set_boolean(RPMOSTREE, "regenerate-initramfs", true);
+        }
+        if let Some(etc) = initramfs.etc.as_deref() {
+            let etc = etc.iter().map(|s| s.as_str());
+            kf_set_string_list(&kf, RPMOSTREE, "initramfs-etc", etc)
+        }
+        if let Some(args) = initramfs.args.as_deref() {
+            let args = args.iter().map(|s| s.as_str());
+            kf_set_string_list(&kf, RPMOSTREE, "initramfs-args", args)
+        }
+    }
+
+    // Custom origin
+    if let Some(custom) = tf.derive.custom.as_ref() {
+        kf.set_string(ORIGIN, "custom-url", custom.url.as_str());
+        if let Some(desc) = custom.description.as_deref() {
+            kf.set_string(ORIGIN, "custom-description", desc);
+        }
+    }
 
-    packages: BTreeSet<String>,
-    packages_local: BTreeMap<String, String>,
-    override_remove: BTreeSet<String>,
-    override_replace_local: BTreeMap<String, String>,
+    if tf.cliwrap.unwrap_or_default() {
+        kf.set_boolean(RPMOSTREE, "ex-cliwrap", true)
+    }
 
-    initramfs_etc: BTreeSet<String>,
-    #[allow(dead_code)]
-    initramfs_args: Vec<String>,
+    if let Some(c) = tf.derive.override_commit.as_deref() {
+        kf.set_string(ORIGIN, "override-commit", c);
+    }
+
```

Comment:
Hmm, though there are lots more options on the compose side which don't make sense on the client side. Should we error out here if the caller is trying to translate those fields? One easy way I guess is to call `origin_to_treefile` and compare it against what we were passed. We'd need to sprinkle `PartialEq` in all the treefile structs.

Reply:
So far...this function is just used to "cross check".  We aren't actually supporting directly processing treefiles on the client - *yet*.  What you're suggesting is really a *triple* check right?  We'd go origin -> treefile -> origin -> treefile and verify that the treefiles are the same?  It makes sense, just uncertain about adding to this PR.

Reply:
Right, mostly concerned with when this will actually be relied upon more strongly. We'll need some way to verify that no unsupported client options were actually passed in. Fine to defer for now.

Reply:
It was actually easy to implement (and a good suggestion) so: added!